component/kafka/component.go

```diff
@@ -0,0 +1,527 @@
+// Package kafka provides kafka component implementation.
+package kafka
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/beatlabs/patron/correlation"
+	patronErrors "github.com/beatlabs/patron/errors"
+	"github.com/beatlabs/patron/internal/validation"
+	"github.com/beatlabs/patron/log"
+	"github.com/beatlabs/patron/trace"
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	propSetMSG        = "property '%s' set for '%s'"
+	consumerComponent = "kafka-consumer"
+	subsystem         = "kafka"
+	messageReceived   = "received"
+	messageProcessed  = "processed"
+	messageErrored    = "errored"
+)
+
+var (
+	consumerErrors           *prometheus.CounterVec
+	topicPartitionOffsetDiff *prometheus.GaugeVec
+	messageStatus            *prometheus.CounterVec
+)
+
+func init() {
+	consumerErrors = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "component",
+			Subsystem: subsystem,
+			Name:      "consumer_errors",
+			Help:      "Consumer errors, classified by consumer name",
+		},
+		[]string{"name"},
+	)
+
+	topicPartitionOffsetDiff = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: subsystem,
+			Name:      "offset_diff",
+			Help:      "Message offset difference with high watermark, classified by topic and partition",
+		},
+		[]string{"group", "topic", "partition"},
+	)
+
+	messageStatus = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "component",
+			Subsystem: subsystem,
+			Name:      "message_status",
+			Help:      "Message status counter (received, processed, errored) classified by topic and partition",
+		}, []string{"status", "group", "topic"},
+	)
+
+	prometheus.MustRegister(
+		consumerErrors,
+		topicPartitionOffsetDiff,
+		messageStatus,
+	)
+}
+
+// consumerErrorsInc increments the number of errors encountered by a specific consumer
+func consumerErrorsInc(name string) {
+	consumerErrors.WithLabelValues(name).Inc()
+}
+
+// topicPartitionOffsetDiffGaugeSet creates a new Gauge that measures partition offsets.
+func topicPartitionOffsetDiffGaugeSet(group, topic string, partition int32, high, offset int64) {
+	topicPartitionOffsetDiff.WithLabelValues(group, topic, strconv.FormatInt(int64(partition), 10)).Set(float64(high - offset))
+}
+
+// messageStatusCountInc increments the messageStatus counter for a certain status.
+func messageStatusCountInc(status, group, topic string) {
+	messageStatus.WithLabelValues(status, group, topic).Inc()
+}
+
+// Builder gathers all required properties in order to construct a batch consumer component
+type Builder struct {
```

Comment:
In the SQS package we actually dumped the builder pattern. The reasons for this is that the functional option pattern is much more flexible, more Go idiomatic and dont require to have 2 structures (Builder and component)

Reply:
Done, implemented the same style `OptionFunc` 👍